### ISSUE TITLE
Move XML stream header processing to respective authenticators

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2772]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)| @ralphm reported|
|-|-|
|Trac ID|trac#2772|
|Type|enhancement|
|Created|2007-08-03 12:50:31Z|

Currently, all stream header processing for Jabber XML streams is done in the `XmlStream` class. However, for checking certain properties, like namespaces, and then sending a stream error in reply, it would be better to leave this to the authenticator.

For example a Jabber service accepting incoming streams, may be able to handle multiple domains. So it needs to inspect the incoming stream header for the addressed domain, and then modify the XmlStream attribute before the outgoing header can be sent. As stream errors also require sending the header first, doing this inside `XmlStream` is not very practical.

Suggested is to move the stream header processing to authenticators. This removes the incoming/outgoing special casing in `XmlStream` and also allows for easier overriding without touching the generic stuff in `XmlStream`.

<details><summary>Searchable metadata</summary>

```
trac-id__2772 2772
type__enhancement enhancement
reporter__ralphm ralphm
priority__highest highest
milestone__Words_0_6 Words-0.6
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1186145431000000 1186145431000000
changetime__1186568287000000 1186568287000000
version__None None
owner__ 
cc__ralphm cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_1"></a>|@ralphm removed owner|
|-|-|

Done in [branches/stream-headers-2772.](branches/stream-headers-2772.)

Also includes support for adding prefix declarations on outgoing stream headers and store them for incoming headers.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to @ralphm|
|-|-|

xmlstream.py:
 * ConnectAuthenticator.streamStarted is missing a docstring (a small one, as it's partly documented in Authenticator.streamStarted)
 * the onFeatures function would be better as a method of the Authenticator, maybe private, unless you have an argument against that.
 * ListenAuthenticator methods are missing docstrings.
 * the job done in ConnectAuthenticator.streamStarted and ListenAuthenticator.streamStarted can probably be factored out a bit, at least the version extraction.
 * the case where IndexError and ValueError are not tested, in both methods. Also, the case where the version attribute is missing in not tested in ListenAuthenticator.streamStarted

That looks mostly good, but I'm a bit worried about backward compatibility. A small note explaining what can be broken would be good.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_4"></a>|@ralphm set owner to @therve|
|-|-|

Thanks for the comments!

I addressed all the comments except for the second bullet. I didn't make onFeatures a separate method because it might obfuscate the call to `initializeStream`. I had thought about using a deferred, but that would be overkill, probably.

I moved the version extraction to the base class `Authenticator` and moved the associated tests, while extending them for `IndexError` and `ValueError`. Actually the implementation was wrong at this, because it didn't properly catch `ValueError`. Thanks for that find.

About backward compatibility: users are not creating their own authenticators, yet, so most of the changes shouldn't cause any trouble. One change might, that is the `thisHost` and `otherHost` attributes on `XmlStream` changing to `thisEntity` and `otherEntity` and having a `JID` as value. But I checked around a bunch of code, and no-one is actually using them so far and the old attributes were fairly recent themselves.

Please review again.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to @ralphm|
|-|-|

I have a test failing now:
```
[ERROR]: twisted.words.test.test_jabberxmlstream.XmlStreamTest.test_authenticator

Traceback (most recent call last):
  File "twisted/trunk/twisted/words/test/test_jabberxmlstream.py", line 377, in test_authenticator
    xs = xmlstream.XmlStream(a)
  File "twisted/trunk/twisted/words/protocols/jabber/xmlstream.py", line 464, in __init__
    authenticator.associateWithStream(self)
exceptions.TypeError: 'list' object is not callable
```

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_6"></a>|@ralphm commented|
|-|-|

Interesting. I cannot reproduce it here. Can you figure out what is happening?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

 * the test is failing with the coverage, you just have to rename the variables (bug in sys.settrace).
 * there are a few ```self.assertEquals(True, foo)``` or ```self.assertEquals(foo, True)```: that would be better handled by assertTrue. There are also some assert_, I would prefer assertTrue too.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_8"></a>|@ralphm set status to `closed`|
|-|-|

(In [20990]) Move XMPP stream header processing to authenticators.

Author: ralphm
Reviewer: therve
Fixes [#2772](https://github.com/twisted/twisted/issues/2772).

This also changes `XmlStream`'s `thisHost` and `otherHost` to `thisEntity` and
`otherEntity` respectively and having them contain `JID`'s as a value.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_9"></a>|@ralphm set status to `reopened`|
|-|-|

Docstring typo.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_10"></a>|@ralphm set status to `closed`|
|-|-|

(In [20997]) Move XMPP stream header processing to authenticators.

Author: ralphm
Reviewer: therve
Fixes [#2772](https://github.com/twisted/twisted/issues/2772).

This also changes `XmlStream`'s `thisHost` and `otherHost` to `thisEntity` and
`otherEntity` respectively and having them contain `JID`'s as a value.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
